src/strategies/flex_window.py

```diff
@@ -84,170 +86,321 @@ def step(self, event_list=[]):
+                    discharge = battery.unload(
+                        self.interval, (total_power/len(batteries)),
+                        target_soc=self.DISCHARGE_LIMIT
+                    )["avg_power"]
+                gc.add_load(b_id, -discharge)
+                timesteps[0]["total_load"] -= discharge
 
-        return charging_stations
-
-    def distribute_peak_shaving(self, timesteps):
+    def distribute_peak_shaving_vehicles(self, timesteps):
 
         gc = list(self.world_state.grid_connectors.values())[0]
-        # order vehicles
+        # get all vehicles that are connected in this step and order vehicles
         vehicles = sorted(
-            [v for v in self.world_state.vehicles.values()
-                if v.connected_charging_station is not None], key=self.sort_key)
+            [
+                v
+                for v in self.world_state.vehicles.values()
+                if v.connected_charging_station is not None
+            ],
+            key=self.sort_key,
+        )
 
         # what happens when all cars are charged with maximum power during charging windows?
         sim_vehicles = deepcopy(vehicles)
         cur_vehicles = sim_vehicles
 
+        # check if battery can be fully charged within time windows in horizon with max power
         cur_time = self.current_time - self.interval
         for ts_info in timesteps:
             cur_time += self.interval
-            cur_vehicles = [v for v in cur_vehicles if v.estimated_time_of_departure > cur_time]
+            # get all vehicle events that depart later than current time
+            cur_vehicles = [
+                v
+                for v in cur_vehicles
+                if (v.estimated_time_of_departure > cur_time)
+                and (v.battery.soc < v.desired_soc)
+            ]
+            # calculate total energy needed
             cur_needed = sum([v.get_energy_needed(full=True) for v in cur_vehicles])
+            # if there is no cars to charge or no energy needed
             if not cur_vehicles or cur_needed < self.EPS:
                 # no cars or no energy need: skip check
                 break
-            self.distribute_power(cur_vehicles, ts_info["power"], cur_needed)
-        charged_in_window = all([v.get_delta_soc() < self.EPS for v in sim_vehicles])
 
+            if ts_info["window"]:  # todo: changed this line
+                self.distribute_power(cur_vehicles, ts_info["power"], cur_needed)
+
+        charged_in_window = all([v.get_delta_soc() < self.EPS for v in sim_vehicles])
+        # apply new soc in this step
         if charged_in_window:
             # can be charged within windows: reset SoC
             for i, v in enumerate(sim_vehicles):
                 v.battery.soc = vehicles[i].battery.soc
+            new_timesteps = [item for item in timesteps if item["window"] is True]
+        else:
+            new_timesteps = [item for item in timesteps if item["window"] is False]
```

Comment:
can be written more elegantly: `new_timesteps = [ts for ts in timesteps if ts["window"] == charged_in_window]`